### PR TITLE
[mlir][math] lower rsqrt to sqrt + fdiv

### PR DESCRIPTION
**Pull request by cferry-AMD**
This commit creates an expansion pattern to lower math.rsqrt(x) into fdiv(1, sqrt(x)).

### CODE REVIEW

**Comment by cferry-AMD** [comment]
This also turns out to be MathToLLVM's current behavior.

**Comment by cferry-AMD** [comment]
For review: @srcarroll @rsuderman @jinchen62 @pashu123, feel free to mention to anyone relevant!

File: mlir/lib/Dialect/Math/Transforms/ExpandPatterns.cpp
Comment (srcarroll) [inline mlir/lib/Dialect/Math/Transforms/ExpandPatterns.cpp:630]: nit: it is preferred to use tablegen designated getter functions for operands and results (otherwise you need to check that you can even index `getOperand(0)`.). so can do `op.getOperand()` instead (see https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Dialect/Math/IR/MathOps.td#L44)

**Review by srcarroll — COMMENTED** [review]
just one nit. otherwise looks good to me. thanks

**Review by pashu123 — COMMENTED** [review]
Please add integration tests here: https://github.com/llvm/llvm-project/blob/main/mlir/test/mlir-cpu-runner/test-expand-math-approx.mlir

File: mlir/lib/Dialect/Math/Transforms/ExpandPatterns.cpp
Comment (pashu123) [inline mlir/lib/Dialect/Math/Transforms/ExpandPatterns.cpp:625]: I think the expansion will work fine for f16, f32, f64. You can enable them.

**Review by pashu123 — COMMENTED** [review]
Some comments.

**Comment by cferry-AMD** [comment]
> out of curiosity, why choose `div` + `sqrt` over `sqrt` + `div`? does it make a difference in terms of accuracy?

I have actually not thought of that other way. Let me have a look... that's a good point!

File: mlir/lib/Dialect/Math/Transforms/ExpandPatterns.cpp
Comment (cferry-AMD) [reply mlir/lib/Dialect/Math/Transforms/ExpandPatterns.cpp:625]: Sure -- @srcarroll, you're right, F16 and F64 are indeed supported, let me add tests for them.

**Comment by cferry-AMD** [comment]
> 
> I have actually not thought of that other way. Let me have a look... that's a good point!

Shallow testing on 1/sqrt(x) for integer values of x did not point out significant accuracy differences. However, the behavior with respect to `-0.0` looks different when hand-writing the two versions into `test-expand-math-approx.mlir`:

```mlir
  %1 = arith.constant -0.0 : f32
  %one = arith.constant 1.0 : f32
  
  // CHECK: -nan
  %2 = arith.divf %one, %1 : f32
  %3 = math.sqrt %2 : f32
  vector.print %3 : f32

  // CHECK: -inf
  %4 = math.sqrt %1 : f32
  %5 = arith.divf %one, %4 : f32
  vector.print %5 : f32
```

* div(1, sqrt(x)) will return -inf,
* sqrt(div(1, x)) will return -nan.

Given that MathToLLVM uses 1/sqrt(x), I'd go the same route to be consistent with it.

File: mlir/lib/Dialect/Math/Transforms/ExpandPatterns.cpp
Comment (cferry-AMD) [reply mlir/lib/Dialect/Math/Transforms/ExpandPatterns.cpp:631]: Added within `test-expand-math-approx.mlir`.

File: mlir/lib/Dialect/Math/Transforms/ExpandPatterns.cpp
Comment (pashu123) [inline mlir/lib/Dialect/Math/Transforms/ExpandPatterns.cpp:630]: can pass the `operand` var directly here.

**Review by pashu123 — COMMENTED** [review]
some nits.

**Comment by cferry-AMD** [comment]
Fixed!

**Review by pashu123 — APPROVED** [review]
LGTM!

**Comment by cferry-AMD** [comment]
I can't merge this (yet), will request write access. Can one of you please do? @pashu123 @srcarroll 
Thanks!

**Pull request merged**
